### PR TITLE
Soundboard

### PR DESCRIPTION
**Pull request by dch-GH**
image

image

Soundboard cooldown is customizable via a project (lobby) setting. Can be disabled entirely.

### CODE REVIEW

File: code/GrubsConfig.cs [hunk 1]
```diff
@@ -111,6 +111,12 @@ public enum TerrainEnvironmentType
 	[ConVar.Replicated( "terrain_environment_type" )]
 	public static TerrainEnvironmentType WorldTerrainEnvironmentType { get; set; } = TerrainEnvironmentType.Sand;
 
+	[ConVar.Replicated( "soundboard_cooldown", Help = "How often (in seconds) players can annoy eachother with sounds." )]
+	public static float SoundboardCooldown { get; set; } = 5;
+
+	[ConVar.Replicated( "soundboard_enabled" )]
+	public static bool SoundboardEnabled { get; set; } = true;
```
Comment (matekdev) [inline code/GrubsConfig.cs:118]: Might be nice to include a sound volume level that the client can change, from 0 -> 1

File: code/GrubsGame.cs [hunk 2]
```diff
@@ -38,6 +40,9 @@ public GrubsGame()
 			Game.SetRandomSeed( (int)(DateTime.Now - DateTime.UnixEpoch).TotalSeconds );
 			PopulatePlayerColors();
 		}
+
+		if ( GrubsConfig.SoundboardEnabled )
+			LoadSoundboard();
```
Comment (matekdev) [inline code/GrubsGame.cs:45]: What happens if `SoundboardEnabled` changes mid-game, are the sounds not loaded?

File: code/GrubsGame.cs [hunk 3]
```diff
@@ -147,6 +152,11 @@ private void PopulatePlayerColors()
 		PlayerColors.Add( Color.FromBytes( 240, 236, 211 ), false ); // Eggshell
 	}
 
+	private void LoadSoundboard()
+	{
+		SoundboardSounds = ResourceLibrary.GetAll<SoundboardSound>().ToList();
```
Comment (matekdev) [inline code/GrubsGame.cs:157]: Could this just be a static list initialized on the variable itself?

`public static List<SoundboardSound> SoundboardSounds = ResourceLibrary.GetAll<SoundboardSound>().ToList();`

File: code/Player/Player.cs [hunk 4]
```diff
@@ -25,6 +25,9 @@ public partial class Player : Entity
 	[Net]
 	public string AvatarClothingData { get; private set; }
 
+	[Net]
+	public RealTimeSince SinceSoundboardPlay { get; set; } = 0;
```
Comment (matekdev) [inline code/Player/Player.cs:29]: `TimeSinceSoundPlayed` or `TimeSinceSoundboardPlayed`

File: code/SoundBoard/SoundBoardSound.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
+namespace Grubs;
+
+[GameResource( "Soundboard Sound", "sb", "A sound to play on the in-game sound board.", Icon = "grid_view" )]
+public class SoundboardSound : GameResource
+{
+	/// <summary>
+	/// The name shown to players in the UI.
+	/// Also used as a parameter in the ServerCmd for playing sounds. 
+	/// </summary>
+	/// <value></value>
```
Comment (matekdev) [inline code/SoundBoard/SoundBoardSound.cs:10]: `/// <value></value>`?

**Review by matekdev — COMMENTED** [review]
1.Can you add the `Tab` key to the InputHints panel
2.Maybe it'd be cool if we showed the voice icon next to the player if they were playing a sound

nice stuff

File: code/GrubsConfig.cs (hunk 1 above)
Comment (dch-GH) [reply code/GrubsConfig.cs:118]: Great idea!

**Comment by dch-GH** [comment]
> 1.Can you add the `Tab` key to the InputHints panel 2.Maybe it'd be cool if we showed the voice icon next to the player if they were playing a sound
> 
> nice stuff

Sure thing

File: code/Soundboard/SoundboardSound.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+	/// The name shown to players in the UI.
+	/// Also used as a parameter in the ServerCmd for playing sounds. 
+	/// </summary>
+	/// <value></value>
+	public string Title { get; set; }
+
+	/// <summary>
+	/// The Emoji to serve as an icon for the sound.
+	/// </summary>
+	/// <value></value>
```
Comment (matekdev) [inline code/Soundboard/SoundboardSound.cs:16]: /// <value></value>

File: code/Soundboard/SoundboardSound.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+namespace Grubs;
+
+[GameResource( "Soundboard Sound", "sb", "A sound to play on the in-game sound board.", Icon = "grid_view" )]
+public class SoundboardSound : GameResource
+{
+	/// <summary>
+	/// The name shown to players in the UI.
+	/// Also used as a parameter in the ServerCmd for playing sounds. 
+	/// </summary>
+	/// <value></value>
```
Comment (matekdev) [inline code/Soundboard/SoundboardSound.cs:10]: /// <value></value>

File: code/UI/PlayerList/PlayerInfo.razor [hunk 8]
```diff
@@ -15,6 +15,7 @@
         <label class="disconnect-icon">link_off</label>
     }
     <label class="voice-icon" style="opacity: @(_isTalking ? 1 : 0)">record_voice_over</label>
+    <label class="soundboard-icon" style="opacity: @(_isUsingSoundboard ? 1 : 0)">graphic_eq</label>
```
Comment (matekdev) [inline code/UI/PlayerList/PlayerInfo.razor:18]: Can you put the icon wiggle as well here

File: code/UI/Soundboard/Soundboard.razor [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+                @onmouseout=@(() => _hoveredSound = null)
+                onclick=@( ()=> 
+				{
+					if( canPlaySound )
+						GrubsGame.PlaySoundboardSound(sound.Title); 
+				})></SoundboardButton>
+		}
+        </div>
+        
+        @* TODO: Add a slider for clientside soundboard volume when sliders get added. *@
```
Comment (matekdev) [inline code/UI/Soundboard/Soundboard.razor:28]: A github issue is probably better than a comment tbh

File: code/UI/Soundboard/Soundboard.razor [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+	public override void Tick()
+	{
+		_isOpen = Input.Down(Grubs.InputAction.Soundboard) && !Inventory.IsOpen;
+        Cursor.Enabled("Soundboard", _isOpen);
+	}
+
+    protected override int BuildHash()
+    {
+        return HashCode.Combine(_isOpen, _hoveredSound?.GetHashCode());
```
Comment (matekdev) [inline code/UI/Soundboard/Soundboard.razor:45]: We need to include the hash code for when playing a sound is unavailable here. 
`var canPlaySound = GrubsConfig.SoundboardEnabled && player.TimeSinceSoundboardPlayed>= GrubsConfig.SoundboardCooldown;`

**Review by matekdev — COMMENTED** [review]
image
Also, any reason we moved this off the picture?

**Comment by dch-GH** [comment]
> image Also, any reason we moved this off the picture?

https://github.com/apetavern/sbox-grubs/commit/665f64b5f0a841b0ac1e07b8ed5472574f299fc2#diff-8b0b3fcc2cd817c095ae8c46dc86a624f534ee96d6e870383ef52f2cbbf611bd

**Comment by matekdev** [comment]
We are going to approach this feature differently in the future, closing for now.

**Pull request closed without merging**